File: data/system/particle.lua
--- Particle effect.
--
-- Lips of Suna is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Lesser General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License, or (at your option) any later version.
--
-- @module system.particle
-- @alias Particle

local Class = require("system/class")

if not Los.program_load_extension("particle") then
	error("loading extension `particle' failed")
end

------------------------------------------------------------------------------

--- Particle effect.
-- @type Particle
local Particle = Class("Particle")

--- Creates a new particle set.
-- @param clss Particle class.
-- @return Particle.
Particle.new = function(clss, model)
	local self = Class.new(clss)
	self.handle = Los.particle_new()
	return self
end

--- Adds a particle effect from a model.
-- @param self Particle.
-- @param model Model.
Particle.add_model = function(self, model)
	Los.particle_add_model(self.handle, model.handle)
end

--- Adds a particle effect from a script.
-- @param self Particle.
-- @param name Script name.
Particle.add_script = function(self, name)
	Los.particle_add_ogre(self.handle, name)
end

--- Instantly clears all the particles.
-- @param self Particle.
Particle.clear = function(self)
	Los.particle_clear(self.handle)
end

--- Enables or disables particle emitting.
-- @param self Particle.
-- @param v True to enable. False otherwise.
Particle.set_emitting = function(self, v)
	Los.particle_set_emitting(self.handle, v)
end

--- Sets the position of the particle set.
-- @param self Particle.
-- @param v Vector.
Particle.set_position = function(self, v)
	Los.particle_set_position(self.handle, v.x, v.y, v.z)
end

--- Sets the render queue of the particle set.
-- @param self Particle.
-- @param name Queue name.
Particle.set_render_queue = function(self, name)
	Los.particle_set_render_queue(self.handle, name)
end

--- Sets the rotation of the particle set.
-- @param self Particle.
-- @param v Quaternion.
Particle.set_rotation = function(self, v)
	Los.particle_set_rotation(self.handle, v.x, v.y, v.z, v.w)
end

--- Sets the size of the particle set.
-- @param self Particle.
-- @param v True to make visible. False otherwise.
Particle.set_visible = function(self, v)
	Los.particle_set_visible(self.handle, v)
end

return Particle

File: src/lipsofsuna/extension/particle/particle.cpp
/* Lips of Suna
 * Copyright© 2007-2013 Lips of Suna development team.
 *
 * Lips of Suna is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Lips of Suna is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Lips of Suna. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * \addtogroup LIExt Extension
 * @{
 * \addtogroup LIExtParticle Particle
 * @{
 */

#include "particle.hpp"
#include "particle-driver.hpp"
#include "particle-driver-ogre.hpp"
#include "particle-driver-model.hpp"

/**
 * \brief Initializes the particle effect.
 * \param render Renderer.
 */
LIExtParticle::LIExtParticle (LIRenRender* render) :
	render(render), emitting(true), looping(true)
{
	render_queue = Ogre::RENDER_QUEUE_MAIN;

	node = render->scene_root->createChildSceneNode ();
	node->setVisible (false);
}

/**
 * \brief Frees the particle effect.
 */
LIExtParticle::~LIExtParticle ()
{
	for (size_t i = 0 ; i < drivers.size() ; i++)
		delete drivers[i];
	render->scene_root->removeAndDestroyChild (node->getName ());
}

/**
 * \brief Adds a particle effect from a model.
 * \param model Model.
 */
void LIExtParticle::add_model (
	const LIMdlModel* model)
{
	LIExtParticleDriverModel* driver = new LIExtParticleDriverModel(this, model);
	driver->set_emitting(emitting);
	driver->set_looping(looping);
	driver->set_render_queue(render_queue);
	drivers.push_back(driver);
}

/**
 * \brief Adds a particle effect from an Ogre script.
 * \param name Script name.
 */
void LIExtParticle::add_ogre (
	const char* name)
{
	LIExtParticleDriverOgre* driver = new LIExtParticleDriverOgre(this, name);
	driver->set_emitting(emitting);
	driver->set_looping(looping);
	driver->set_render_queue(render_queue);
	drivers.push_back(driver);
}

/**
 * \brief Instantly removes all the particles.
 */
void LIExtParticle::clear ()
{
	for (size_t i = 0 ; i < drivers.size() ; i++)
		delete drivers[i];
	drivers.clear ();
}

/**
 * \brief Updates the particle effect.
 * \param secs Seconds since the last update.
 */
void LIExtParticle::update (
	float secs)
{
	for (size_t i = 0 ; i < drivers.size() ; i++)
		drivers[i]->update (secs);
}

/**
 * \brief Enables or disables particle emission.
 * \param value True to enable. False otherwise.
 */
void LIExtParticle::set_emitting (
	bool value)
{
	emitting = value;
	for (size_t i = 0 ; i < drivers.size() ; i++)
		drivers[i]->set_emitting (value);
}

/**
 * \brief Enables or disables looping.
 * \param value True to enable. False otherwise.
 */
void LIExtParticle::set_looping (
	bool value)
{
	looping = value;
	for (size_t i = 0 ; i < drivers.size() ; i++)
		drivers[i]->set_looping (value);
}

/**
 * \brief Sets the render queue of the particle effect.
 * \param value Queue name.
 */
void LIExtParticle::set_render_queue (
	const char* value)
{
	if (!strcmp (value, "background"))
		render_queue = Ogre::RENDER_QUEUE_BACKGROUND;
	else if (!strcmp (value, "skies_early"))
		render_queue = Ogre::RENDER_QUEUE_SKIES_EARLY;
	else if (!strcmp (value, "1"))
		render_queue = Ogre::RENDER_QUEUE_1;
	else if (!strcmp (value, "2"))
		render_queue = Ogre::RENDER_QUEUE_2;
	else if (!strcmp (value, "3"))
		render_queue = Ogre::RENDER_QUEUE_3;
	else if (!strcmp (value, "4"))
		render_queue = Ogre::RENDER_QUEUE_4;
	else if (!strcmp (value, "main"))
		render_queue = Ogre::RENDER_QUEUE_MAIN;
	else if (!strcmp (value, "6"))
		render_queue = Ogre::RENDER_QUEUE_6;
	else if (!strcmp (value, "7"))
		render_queue = Ogre::RENDER_QUEUE_7;
	else if (!strcmp (value, "8"))
		render_queue = Ogre::RENDER_QUEUE_8;
	else if (!strcmp (value, "9"))
		render_queue = Ogre::RENDER_QUEUE_9;
	else if (!strcmp (value, "skies_late"))
		render_queue = Ogre::RENDER_QUEUE_SKIES_LATE;
	else if (!strcmp (value, "overlay"))
		render_queue = Ogre::RENDER_QUEUE_OVERLAY;

	for (size_t i = 0 ; i < drivers.size() ; i++)
		drivers[i]->set_render_queue (render_queue);
}

/**
 * \brief Frees the particle effect.
 * \param self Particle effect.
 */
void liext_particle_free (
	LIExtParticle* self)
{
	delete self;
}

/** @} */
/** @} */
